### PR TITLE
JAMES-3693 RedisHealthCheck

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package org.apache.james.rate.limiter.redis
```
Comment (Arsnael) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:1]: license missing

File: server/mailet/rate-limiter-redis/src/test/java/org/apache/james/rate/limiter/RedisHealthCheckTest.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package org.apache.james.rate.limiter
```
Comment (Arsnael) [inline server/mailet/rate-limiter-redis/src/test/java/org/apache/james/rate/limiter/RedisHealthCheckTest.scala:1]: license missing

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import reactor.core.scala.publisher.SMono
+
+import scala.concurrent.duration.FiniteDuration
+
+sealed class RedisHealthCheckKey() extends RateLimitingKey {
+  override def asString(): String = "RedisHealthCheck"
+}
+
+class RedisHealthCheck @Inject()(redisRateLimiterFactory: RedisRateLimiterFactory) extends HealthCheck {
+  private val COMPONENT_NAME: ComponentName = new ComponentName("Redis")
```
Comment (vttranlina) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:20]: it is not a static variable, we should use camelCase

**Review by Arsnael — APPROVED** [review]

**Review by hungphan227 — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.SMono
+
+import scala.concurrent.duration.FiniteDuration
+
+sealed class RedisHealthCheckKey() extends RateLimitingKey {
+  override def asString(): String = "RedisHealthCheck"
+}
+
+class RedisHealthCheck @Inject()(redisRateLimiterFactory: RedisRateLimiterFactory) extends HealthCheck {
```
Comment (chibenwa) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:38]: IMO basing a "generic" component like a healthcheck on an applicative feature is brittle. 

In a redis health check I'd expect only the Redis layer to be tested.

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  private val redisHealthcheckRateLimiter = redisRateLimiterFactory.withSpecification(flexibleRateLimitRule, None)
+  private val healthcheckTimeout = new FiniteDuration(3, TimeUnit.SECONDS)
+
+  override def componentName(): ComponentName = redisComponent
+
+  override def check(): Publisher[Result] =
+    SMono.fromPublisher(redisHealthcheckRateLimiter.rateLimit(redisHealthcheckKey, 1))
+      .timeout(healthcheckTimeout)
+      .`then`(SMono.just(Result.healthy(redisComponent)))
+      .onErrorResume(e => SMono.just(Result.unhealthy(redisComponent, "Can not connect to Redis.", e)))
```
Comment (chibenwa) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:51]: degraded -> redis shall never be critical component

**Review by Arsnael — APPROVED** [review]

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 6]
```diff
@@ -20,33 +20,43 @@
...
+
+      SMono.fromPublisher(Mono.using(() => RedisClusterClient.create(redisUris),
+        (redisClusterClient: RedisClusterClient) => Mono.fromCallable(() => redisClusterClient.connect()),
+        (redisClusterClient: RedisClusterClient) => redisClusterClient.close()))
+    } else {
+      val redisUri: RedisURI = redisConfiguration.redisURI.value.last
+      redisUri.setTimeout(healthcheckTimeout)
+
+      SMono.fromPublisher(Mono.using(() => RedisClient.create(redisUri),
+        (redisClient: RedisClient) => Mono.fromCallable(() => redisClient.connect()),
```
Comment (chibenwa) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:59]: Is connect blocking? If yes, is there a way to do in on a non blocking fashion? 
If impossible please do this on boundedElastic.

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 7]
```diff
@@ -47,16 +49,18 @@ class RedisHealthCheck @Inject()(redisConfiguration: RedisRateLimiterConfigurati
...
       val redisUris = redisConfiguration.redisURI.value.asJava
       redisUris.forEach(redisUri => redisUri.setTimeout(healthcheckTimeout))
+      val redisClusterClient = RedisClusterClient.create(redisUris)
 
-      SMono.fromPublisher(Mono.using(() => RedisClusterClient.create(redisUris),
-        (redisClusterClient: RedisClusterClient) => Mono.fromCallable(() => redisClusterClient.connect()),
-        (redisClusterClient: RedisClusterClient) => redisClusterClient.close()))
+      SMono.fromFuture(redisClusterClient.connectAsync(StringCodec.UTF8).asScala)
+        .doOnNext(redisConnection => redisConnection.closeAsync())
+        .doOnTerminate(() => redisClusterClient.close())
```
Comment (chibenwa) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:56]: That one is still blocking?

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala [hunk 8]
```diff
@@ -47,16 +49,18 @@ class RedisHealthCheck @Inject()(redisConfiguration: RedisRateLimiterConfigurati
...
       val redisUri: RedisURI = redisConfiguration.redisURI.value.last
       redisUri.setTimeout(healthcheckTimeout)
+      val redisClient = RedisClient.create(redisUri)
 
-      SMono.fromPublisher(Mono.using(() => RedisClient.create(redisUri),
-        (redisClient: RedisClient) => Mono.fromCallable(() => redisClient.connect()),
-        (redisClient: RedisClient) => redisClient.close()))
+      SMono.fromFuture(redisClient.connectAsync(StringCodec.UTF8, redisUri).asScala)
+        .doOnNext(redisConnection => redisConnection.closeAsync())
+        .doOnTerminate(() => redisClient.close())
```
Comment (chibenwa) [inline server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:64]: Idem?

File: server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala (hunk 7 above)
Comment (quantranhong1999) [reply server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:56]: Which ones do you mean?

`redisClusterClient.connectAsync `+ `redisConnection.closeAsync` -> non blocking
`redisClusterClient.close()` -> release Java objects, should be fast enough.

Comment (quantranhong1999) [reply server/mailet/rate-limiter-redis/src/main/scala/org/apache/james/rate/limiter/redis/RedisHealthCheck.scala:56]: Ah actually there is a `shutdownAsync` for this (async close both connection + java objects)

**Review by Arsnael — APPROVED** [review]

**Pull request merged**